chore: initial package practices (#117)
defenseunicorns/uds-common: docs/uds-package-practices.md

Reviewer: Is this implemented on either UDS-Core or SWF so examples can be linked here?

# UDS Package Practices

This document describes the practices that a UDS package **must**, **should**, and **may** follow to be considered officially supported as a UDS package and to be able to display the "Made for UDS" badge on its GitHub repository README.md file:

[<img alt="Made for UDS" src="./made-for-uds.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core)

> [!NOTE]
> This badge should link to the `uds-core` repo and should match the height of any other badges (i.e. `20px`).

## Integrations

> Below are the minimal services that a UDS package will integrate with at some level and they:

Reviewer: ```suggestion
Below are the minimal services that a UDS package **must** integrate with:
```

Author: I had this originally :sweat_smile:

### Istio

> - **Must** define external interfaces under the `expose` key **_(if applicable - i.e. has a web interface)_**.

Reviewer 2: may, since _if applicable_ ?

Author: Was trying to make it less ambiguous ( didn't want people making their own VSs for example) - we could preface it with the `if`?

Author: Actually just reworded them to be a bit clearer

- **Must** deploy and operate successfully with Istio injection enabled in the namespace.
- **Should** avoid workarounds such as disabling strict mTLS peer authentication.

### Network Policies

- **Must** define network policies under the `allow` key as required.
- **Should** minimize network policies to specific selectors needed for Ingress/Egress traffic.
- **May** template network policy keys to provide flexibility for delivery customers to configure.

### Keycloak

> - **Must** create a client through the `sso` key and use that client for any provided user login.

Author: Tried covering that with `any provided user login` - i.e. if the tool doesn't provide a way for a user to login (like GLR) than there would be no keycloak integration.  Will think through ways to reword though...

> - **Must** create a client through the `sso` key and use that client for any provided user login.

Reviewer 3: ```suggestion
- **Must** protect any external endpoints with a client through the `sso` key and use that client for any provided user login.
```

Is this too aggressive? I think in general you'd always want keycloak auth, even if it's with Authservice (want to be careful not to say "if app implements SSO"). There may still be some public external endpoints though...static docs stuff?

> - **Must** create a client through the `sso` key and use that client for any provided user login.

Author: refined the verbiage a bit

> - **Must** create a client through the `sso` key and use that client for any provided user login.

Reviewer 3: I think this is good from my perspective - once Authservice rolls out it might make sense to clarify/restrict this even more to expect authservice integration in some cases.

- **Should** consider security options during implementation to provide the most secure default possible.
- **Should** name the client `<App> Login` (i.e. `Mattermost Login`) to provide UX consistency.
- **May** template Keycloak fields to provide flexibility for delivery customers to configure.

## Structure

Packages also follow structural guidelines to ensure consistency and flexibility for configuration, they:

- **Should** expose configuration through a chart (ideally in a `chart` or `charts` directory).

>   > This allows UDS bundles to override configuration with Helm overrides.

Reviewer: ```suggestion
  > This allows UDS bundles to override configuration with Helm overrides. Enabling downstream teams full control of their bundle configurations.
```

- **Should** implement or allow for multiple flavors (ideally with common definitions in a `common` directory)
  > This allows for different images or configurations to be delivered consistently to customers.

## Testing

> A UDS Package will also have testing and quality checks to ensure that updates / changes to them will flow smoothly over time.  Packages:

Reviewer: ```suggestion
A UDS Package will also have testing and quality checks to ensure that updates / changes to them result in minimal churn.  Packages:
```

- **Must** implement Journey Testing to cover the basic user flows and features of the application, especially where an application interacts with an external service / interface.

- **Must** implement Upgrade Testing to ensure that the current development package works when deployed over the previously released one.

> - **Should** lint their configurations with appropriate tooling such as `yamllint` and `zarf dev lint`.

Reviewer: link to `yamllint` and `zarf dev lint` CLI's / command pages?

## Maintenance

To help maintain a UDS Package, it:

- **Must** have a dependency management bot (such as renovate) configured to open PRs to update core package and support dependencies.

> - **Must** release its package to the `ghcr.io/defenseunicorns/packages/<group>` namespace as the application's name (i.e. `ghcr.io/defenseunicorns/packages/uds/mattermost`)

Reviewer: Will there be nesting restrictions?

ie, restrict to `<group>/<name>`, or is `<group>/<sub-group>/<name>` allowed?

Author: I think that is fine to leave ambiguous for now - mostly wanted to say they should have a grouping under packages and that packages themselves should have a simple name.

[<img alt="Made for UDS" src="./made-for-uds.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core)

> [!NOTE]
> This badge should link to the `uds-core` repo and should match the height of any other badges (i.e. `20px`).

## Integrations

Below are the minimal services that a UDS package **must** integrate with, using the [UDS `Package` custom resource](https://github.com/defenseunicorns/uds-core/blob/main/src/pepr/operator/README.md#example-uds-package-cr):

### Istio

- **Must** define any external interfaces under the `expose` key.
- **Must** deploy and operate successfully with Istio injection enabled in the namespace.
- **Should** avoid workarounds such as disabling strict mTLS peer authentication.

### Network Policies

- **Must** define network policies under the `allow` key as required.
- **Should** minimize network policies to specific selectors needed for Ingress/Egress traffic.
- **May** template network policy keys to provide flexibility for delivery customers to configure.

### Keycloak

> - **Must** use and create a Keycloak client through the `sso` key if the application provides a user login.

Reviewer 4: Should we also give some guidance about the client id and the secret name? More of a Developer/Admin experience thing. I'd say 'uds-core-<app>' for the client id and '<app>-sso' in the <app> namespace for the secret is the common pattern right now.

Author: We could though we'd need to make it work for leapfrog too (which doesn't use the `uds` brand).  IMO `uds-core-mattermost` is also a misnomer since mattermost is not in uds-core.  should be `uds-swf-mattermost` most likely.

Reviewer 3: Chiming in my 2cents on the client id:
- I think the `uds-` prefix is fine, this is all in the context of things "made for uds" so I think we're okay enforcing that if we want
- It should definitely have the "app name/identifier" (i.e. mattermost, grafana, sonarqube, ...)
- Where applicable it may make sense to add the "app group" (i.e. core/swf/...)

I don't know if it makes sense to be prescriptive on the secret name - for reference the `neuvector` chart only allows you to mount secrets that match a specific name (`neuvector-init`) and does not expose values to point at a different secret. This is probably not common but I would hesitate to strictly enforce this one.

# UDS Package Practices

This document describes the practices that a UDS package **must**, **should** and **may** follow to be considered officially supported as a UDS package and to be able to display the "Made for UDS" badge on its GitHub repository README.md file :

[<img alt="Made for UDS" src="./made-for-uds.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core)

> [!NOTE]
> This badge should link to the `uds-core` repo and should match the height of any other badges (i.e. `20px`).

> [!TIP]

> > See [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) for definitions of **must**, **should** and **may**

Reviewer 4: ```suggestion
> This document follows [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) for definitions of requirement levels (e.g. **must**, **should** and **may**)
```

Reviewer 4: ^ wording suggestion to make it more inclusive if we ever reference any other levels like **must not**

## Integrations

Below are the minimal services that a UDS package **must** integrate with, using the [UDS `Package` custom resource](https://github.com/defenseunicorns/uds-core/blob/main/src/pepr/operator/README.md#example-uds-package-cr).  These integrations **must** be declarative and automated without requiring click-ops from the person deploying the package.

### Istio

- **Must** define any external interfaces under the `expose` key.
- **Must** deploy and operate successfully with Istio injection enabled in the namespace.
- **Should** avoid workarounds such as disabling strict mTLS peer authentication.

### Network Policies

- **Must** define network policies under the `allow` key as required.
- **Should** minimize network policies to specific selectors needed for Ingress/Egress traffic.
- **May** template network policy keys to provide flexibility for delivery customers to configure.

### Keycloak

- **Must** use and create a Keycloak client through the `sso` key if the application provides a user login.

> - **Should** consider security options during implementation to provide the most secure default possible.

Reviewer 5: Do we want to call out SAML specifically here instead of OIDC?

- **Must** minimize the scope and number of the exemptions to only what is absolutely required by the application
- **Must** have documented rationale for any exemptions present

## Structure

Packages also follow structural guidelines to ensure consistency and flexibility for configuration, they:

- **Should** expose all configuration (`uds.dev` CRs, additional `Secrets`/`ConfigMaps`, etc) through a Helm chart (ideally in a `chart` or `charts` directory).
  > This allows UDS bundles to override configuration with Helm overrides and enables downstream teams to fully control their bundle configurations.

- **Should** implement or allow for multiple flavors (ideally with common definitions in a `common` directory)
  > This allows for different images or configurations to be delivered consistently to customers.

## Testing

A UDS Package will also have testing and quality checks to ensure that updates / changes to them result in minimal churn.  Packages:

> - **Must** implement Journey Testing to cover the basic user flows and features of the application, especially where an application interacts with an external service / interface.

Reviewer 5: Are there any recommended technologies/tools to implement this type of testing?
